File: src/lib/geo.ts
/**
 * Coordinate primitives (ADR-0022) — `Point` and the haversine math outlive any one Path: it's
 * also the type of a clustering centroid (`metroCluster.ts`) and a discovery corridor endpoint.
 * Split out of the dissolved `travelCost.ts` so nothing here pulls in Path/provider machinery to
 * do plain trig.
 */

export interface Point {
  lat: number;
  lng: number;
}

/** A location is treated as not-yet-geocoded when its coordinates default to (0,0) — the shape
 * `toInput()` (optimize.ts) produces before a Location has real lat/lng. Shared by every caller
 * that needs to exclude these from distance-lookup construction or anchor selection. */
export function hasValidCoords(l: Point): boolean {
  return l.lat !== 0 || l.lng !== 0;
}

const EARTH_RADIUS_M = 6_371_000;

function toRad(deg: number): number {
  return (deg * Math.PI) / 180;
}

/** Straight-line distance in meters — the single haversine implementation shared by the
 * path-provider default, `metroCluster.ts`'s clustering, discovery (discovery.ts, places.ts), and
 * transit-graph ingest/query (transitGraphIngest.ts, transitGraph.ts). `haversineKm` (the pre-VROOM
 * optimizer's k-means centroid math, its only caller) was deleted with `optimizer.ts` (ADR-0023 §9)
 * — every remaining caller wants meters. */
export function haversineMeters(a: Point, b: Point): number {
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const sinDLat = Math.sin(dLat / 2);
  const sinDLng = Math.sin(dLng / 2);
  const x =
    sinDLat * sinDLat +
    Math.cos(toRad(a.lat)) * Math.cos(toRad(b.lat)) * sinDLng * sinDLng;
  return EARTH_RADIUS_M * 2 * Math.atan2(Math.sqrt(x), Math.sqrt(1 - x));
}

File: swift/TripKrakenKit/Sources/TripKrakenKit/Geo.swift
import Foundation

/// Coordinate primitives (ADR-0022) — `Point` and the haversine math outlive any one `Path`: it's
/// also the type of a clustering centroid and a discovery corridor endpoint. Ported from
/// `src/lib/geo.ts`.
public struct Point: Sendable, Hashable, Codable {
    public var lat: Double
    public var lng: Double

    public init(lat: Double, lng: Double) {
        self.lat = lat
        self.lng = lng
    }
}

extension Point {
    /// A location is treated as not-yet-geocoded when its coordinates default to (0, 0). Shared by
    /// every caller that needs to exclude these from distance-lookup construction or anchor
    /// selection.
    public var hasValidCoords: Bool {
        lat != 0 || lng != 0
    }
}

private let earthRadiusMeters = 6_371_000.0

private func toRadians(_ degrees: Double) -> Double {
    degrees * .pi / 180
}

/// Straight-line distance in meters — the single haversine implementation every distance-estimating
/// caller (clustering, discovery, transit-graph ingest/query) shares.
public func haversineMeters(_ a: Point, _ b: Point) -> Double {
    let dLat = toRadians(b.lat - a.lat)
    let dLng = toRadians(b.lng - a.lng)
    let sinDLat = sin(dLat / 2)
    let sinDLng = sin(dLng / 2)
    let x = sinDLat * sinDLat
        + cos(toRadians(a.lat)) * cos(toRadians(b.lat)) * sinDLng * sinDLng
    return earthRadiusMeters * 2 * atan2(x.squareRoot(), (1 - x).squareRoot())
}
